kubemarine/kubernetes/components.py

```diff
@@ -0,0 +1,1051 @@
+                           edit_functions: Dict[str, Callable[[dict], dict]] = None,
+                           force_restart: bool = False) -> None:
+    """
+    Reconfigure the specified `components` on `group` of nodes.
+
+    :param group: nodes to reconfigure components on
+    :param components: List of control plane components or `kube-proxy`, or `kubelet` to reconfigure.
+    :param edit_functions: Callables that edit the specified kubeadm-managed ConfigMaps in a custom way.
+                           The ConfigMaps are fetched from the first control plane,
+                           instead of being generated.
+                           This implies necessity of working API server.
+    :param force_restart: Restart the given `components` even if nothing has changed in their configuration.
+    """
+    if edit_functions is None:
+        edit_functions = {}
+
+    cluster: KubernetesCluster = group.cluster
+    logger = cluster.log
+    control_plane = cluster.nodes['control-plane'].get_first_member()
+
+    timestamp = utils.get_current_timestamp_formatted()
+    backup_dir = '/etc/kubernetes/tmp/kubemarine-backup-' + timestamp
+    logger.debug(f"Using backup directory {backup_dir}")
+
+    logger.debug("Uploading cluster config to first control plane...")
+
+    kubeadm_config = KubeadmConfig(cluster)
+    for configmap, func in edit_functions.items():
+        kubeadm_config.load(configmap, control_plane, func)
+
+    # This configuration will be used for `kubeadm init phase upload-config` phase.
+    # InitConfiguration is necessary to specify patches directory.
+    # Use patches from the first control plane, to make it the same as during regular installation.
+    # Patches are used to upload KubeletConfiguration for some reason (which is likely a gap)
+    # https://github.com/kubernetes/kubernetes/issues/123090
+    upload_config = '/etc/kubernetes/upload-config.yaml'
+    _upload_config(cluster, control_plane, kubeadm_config, upload_config)
+
+    control_plane_components = [
+        c for c in components
+        if c in ('kube-apiserver/cert-sans', 'kube-apiserver', 'kube-scheduler', 'kube-controller-manager', 'etcd')]
+
+    # Firstly reconfigure components that do not require working API server
+    if control_plane_components:
+        if _reconfigure_control_plane_components(cluster, group, control_plane_components, force_restart,
+                                                 kubeadm_config, backup_dir):
+            _update_configmap(cluster, control_plane, 'kubeadm-config',
+                              _configmap_init_phase_uploader('kubeadm-config', upload_config),
+                              backup_dir)
+
+    node_components = [c for c in components if c in ('kubelet', 'kube-proxy')]
+
+    if node_components:
+        _reconfigure_node_components(cluster, group, node_components, force_restart, control_plane,
+                                     upload_config, kubeadm_config, backup_dir)
```

Comment:
Not clear why we do 2 cycles on nodes:

- first cycle updates only manifests
- second cycle updates kubelet/kube-proxy

Why not do one cycle which updates both in the same order?

Reply:
Changed to one cycle in https://github.com/Netcracker/KubeMarine/pull/592/commits/9be4c4801fafec8b3eb623ee6003eeea7b1b225a